SIPS/sip-58.md

```diff
@@ -0,0 +1,86 @@
+
+## Simple Summary
+
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the SIP.-->
+
+Emit list of individual fee reclamations and rebates during exchange settlement for Dapps and Synthetix Exchange.
+
+## Abstract
+
+<!--A short (~200 word) description of the technical issue being addressed.-->
+
+In order to help users track fees claimed or rebated from individual transactions, emit events during settlement that match each individual exchange being settled with the amount reclaimed or rebated.
+
+Upgrade the Exchanger `settle()` function to emit individual fee reclaimation / rebate amounts for each trade.
+
+Add an internal function `_settlementsOwing` that will emit an event `ExchangeEntriesReclaim` and `ExchangeEntriesRebate` when `Exchanger.settle()` is invoked.
+
+## Motivation
+
+<!--The motivation is critical for SIPs that want to change Synthetix. It should clearly explain why the existing protocol specification is inadequate to address the problem that the SIP solves. SIP submissions without sufficient motivation may be rejected outright.-->
+
+We want to display on Synthetix exchange the corresponding fee reclamation amounts (if any) for each individual exchange made when exchange settlement is invoked.
+
+Currently invoking `settle()` will only emit one event, if any, with the total aggregated sum of any fee reclamation or rebate amounts. This makes it difficult and complex for users trying to determine the dividual settlement rates on previous trades they've made.
+
+It is important that traders can see on each trade the fee reclamation and rebates for calculating trading profits and loses based on the amounts and effective price they recieved on each individual trade.
+
+## Specification
+
+<!--The technical specification should describe the syntax and semantics of any new feature.-->
+
+### Exchanger.settle ###
+
+Add an internal function `_settlementsOwing` that will emit an event `ExchangeEntryReclaim` for each reclaim and `ExchangeEntryRebate` for each rebate when `Exchanger.settle()` is invoked.
+
+`Exchanger.settle()` will use `_settlementsOwing` to calculate the fee reclamation amounts.
+
+**Event**
+
+Emit an event `ExchangeEntryReclaim` or `ExchangeEntryRebate` for each exchangeEntry within `_settlementsOwing` when `Exchanger.settle()` is invoked.
+
+```solidity
+event ExchangeEntryReclaim(address indexed from, bytes32 src, uint amount, bytes32 dest, uint reclaimAmount, uint srcRoundIdAtPeriodEnd, uint destRoundIdAtPeriodEnd, uint timestamp);
```

Comment:
hmmmm... i wonder if we should also emit an event when the exchange entry is first added as well? And emit the rates of the `src` and `dest` currencies, then emit the rates of `src` and `dest` at period end (rather than the `roundIds` which are harder to correlate)? It's a bit more work cause we'd need to fetch the rates from `ExchangeRates` but if the goal is to make reporting easier... (Keep in mind that fetching state past 128 blocks back can't be done without an archive node, and our dApps will often want to inform users what the difference in rate was). cc @clementbalestrat